crates/project_panel/src/project_panel.rs

```diff
@@ -1396,6 +1528,40 @@ impl ProjectPanel {
         }
     }
 
+    fn calculate_depth_and_difference(
+        entry: &Entry,
+        visible_worktree_entries: &Vec<Entry>,
+    ) -> (usize, usize) {
+        let entry_path_components_count = entry.path.components().count();
+        let (depth, difference) = entry
+            .path
+            .ancestors()
+            .skip(1) // Skip the entry itself
+            .find_map(|ancestor| {
+                visible_worktree_entries
```

Comment:
The performance of this section worries me, as it's happening on every render, potentially iterating over all of the visible entries for each component of each entry.

Reply:
Hello Mikayla, I have added a `HashSet` in order to improve lookup time, but this solution is still not significantly faster, and I could not come up with a solution to improve it. If you have any better ideas on how to achieve this, please let me know.

Reply:
Ideally you'd have a structure that mirrors the `TreeMap` in the snapshot and contains a flag indicating whether or not each entry was folded. That way you could simply iterate over both *simultaneously* and always have the flag for the current element available with no costly lookups. Question then is how/where to construct/update said map.

If this was part of the worktree snapshot then it could even be directly part of the Entry, but I don't have a good mental model of what level of separation of concerns is desired here...

Reply:
This seems right, and it seems like we'd want to store that map on the ProjectPanel, as it's purely visual information.

Reply:
Does that give you some ideas for what to do next @ABckh?

Reply:
In the first pull request, where I added HashMap to store state for verifying whether directories should be folded or not, @maxbrunsfeld suggested eliminating the storage of this type of state. It seems to me that @jansol's approach is quite similar and involves adding state storage. So, just to double-check, is this the right way to proceed with this feature?

Reply:
Are you supposed to be able to manually fold/unfold directories or not? If so, then that is by definition state that needs to be stored *somewhere*, no?

If manual folding/unfolding is not wanted then yeah, the whole state thing is unnecessary.